modules/docs/markdown/06-guides/smithy-transformations.md

```diff
@@ -0,0 +1,124 @@
+
+## Implement the Transformer
+
+Now we will walk through the steps of actually implementing the Smithy model transformer. The `smithy-build` Java library provides an interface for these called `ProjectionTransformer`. Here is an example of a transformer that will remove the members marked with the `removeBeforeCodegen` trait as discussed above.
+
+```scala
+final class RemoveBeforeCodegenTransformation extends ProjectionTransformer {
+  def getName() = {
+    "RemoveBeforeCodegenTransformation"
+  }
+
+  def transform(x: TransformContext) = {
+    val toRemove = x
+      .getModel()
+      .getShapesWithTrait(new ToShapeId() {
+        override def toShapeId(): ShapeId =
+          ShapeId.from("preprocessed#removeBeforeCodegen")
+      })
+    x.getTransformer().removeShapes(x.getModel(), toRemove)
+  }
+}
+```
+
+Here we create a `ProjectionTransformer` named `RemoveBeforeCodegenTransformation`. Inside the `transform` method we remove all shapes that are marked with the `removeBeforeCodegen` trait and return the resultant model.
+
+## Register the Transformer
+
+We need to register the Transformer so that the Smithy tooling will be able to find it when it does a lookup for all `ProjectionTransformer` implementations. We do this by creating a file at `src/main/resources/META-INF/services/` called `software.amazon.smithy.build.ProjectionTransformer`. Inside of this file, we need to place the fully qualified classpath that leads to our transformer (e.g. `preprocessed.transformations.RemoveBeforeCodegenTransformation`).
+
+## Wire up in SBT
+
+In our `build.sbt` file we will create a new project called `transformers` that looks as follows:
+
+```scala
+lazy val transformers = (project in file("transformers"))
+  .settings(
+    name := "transformers",
+    libraryDependencies += "software.amazon.smithy" % "smithy-build" % smithy4s.codegen.BuildInfo.smithyVersion
+  )
+```
+
+This project is where our `RemoveBeforeCodegenTransformation` will live. As such, it requires the `smithy-build` dependency so it can access the `ProjectionTransformer` interface.
+
+Now, in our project that is using the Smithy4s SBT plugin (`.enablePlugins(Smithy4sCodegenPlugin)`) we need to add the following settings:
+
+```scala
+...
+.enablePlugins(Smithy4sCodegenPlugin)
+.settings(
+  ...
+  // transformers
+  Compile / smithy4sModelTransformers += "RemoveBeforeCodegenTransformation",
+  Compile / smithy4sAllDependenciesAsJars += (transformers / Compile / packageBin).value
+)
+```
+
+Here we are telling the smithy4s SBT plugin which transformers should be applied prior to code generation. In this case, "RemoveBeforeCodegenTransformation" is the only one. From there, we are telling the plugin that it should use the transformers SBT project we created above as a dependency. Adding it this way will make sure that your transformer is NOT added as a dependency to anything other than the Smithy4s plugin.
+
+## Directory Structure
```

Comment:
before this, as a user I'd like to show the actual outcome of the transformation. Maybe we can show a trimmed snippet of the generated case class?

the snippet would be sth like

```scala
// note the lack of the `id` field
case class MyStruct(name: String)
//...
```